File: Sources/Automerge/path.swift
import struct AutomergeUniffi.PathElement
import enum AutomergeUniffi.Prop

typealias FfiPathElem = AutomergeUniffi.PathElement
typealias FfiProp = AutomergeUniffi.Prop

/// A component of the path to an object within a document.
///
/// A path constructed of `PathElement` instances represents the sequence of navigating through a hierarchical structure
/// of objects within an Automerge document.
/// The base of this tree structure is  ``ObjId/ROOT``.
public struct PathElement: Equatable {
    let prop: Prop
    let obj: ObjId

    /// Creates a new path element.
    /// - Parameters:
    ///   - obj: The object Id of the path element.
    ///   - prop: The property on the object, either a key on a map, or index position in a list.
    public init(obj: ObjId, prop: Prop) {
        self.prop = prop
        self.obj = obj
    }

    static func fromFfi(_ ffiElem: FfiPathElem) -> Self {
        Self(
            obj: ObjId(bytes: ffiElem.obj),
            prop: Prop.fromFfi(ffiElem.prop)
        )
    }
}

/// A type that represents a property on an object within an Automerge document.
///
/// The property is either a ``Prop/Key(_:)``, in the from of a `String` to a map,
/// or a ``Prop/Index(_:)`` with the index position represented as a 64-bit unsigned integer.
public enum Prop: Equatable {
    /// A property in a map.
    case Key(String)
    /// An index into a sequence.
    case Index(UInt64)

    static func fromFfi(_ ffi: FfiProp) -> Self {
        switch ffi {
        case let .index(value):
            return .Index(value)
        case let .key(value):
            return .Key(value)
        }
    }
}

File: rust/src/path.rs
use automerge as am;

use super::ObjId;

pub enum Prop {
    Key { value: String },
    Index { value: u64 },
}

impl From<am::Prop> for Prop {
    fn from(value: am::Prop) -> Self {
        match value {
            am::Prop::Map(k) => Prop::Key { value: k },
            am::Prop::Seq(i) => Prop::Index { value: i as u64 },
        }
    }
}

impl From<Prop> for am::Prop {
    fn from(value: Prop) -> Self {
        match value {
            Prop::Key { value } => am::Prop::Map(value),
            Prop::Index { value } => am::Prop::Seq(value as usize),
        }
    }
}

pub struct PathElement {
    pub prop: Prop,
    pub obj: ObjId,
}

impl PathElement {
    pub fn new<P: Into<Prop>, O: Into<ObjId>>(prop: P, obj: O) -> Self {
        Self {
            obj: obj.into(),
            prop: prop.into(),
        }
    }
}
